### PR TITLE
Improve react-components build system

### PR DESCRIPTION
**Pull request by pjanik**
https://www.pivotaltracker.com/story/show/187689236

This PR should have a reviewable diff, as GitHub seems not to show all the moved files. However, if it doesn't work for you, you could review it commit by commit, skipping the [chore: move /react-components dir to /rails/react-components](https://github.com/concord-consortium/rigse/commit/a6ddecbec68c5648fc31f8183731f6dd471c8750) commit.

The main idea was to remove the built react-component.js file from the git repository and update the webpack configuration so webpack builds directly into the location in the Rails assets pipeline structure. This way, developers won't have to remember to build react-components before each commit, avoiding merge conflicts, and we can use `webpack --watch` for very fast incremental builds (10s -> ~1s or less).

@dougmartin , @scytacki, I am not very confident with the Docker setup, so I'd focus especially on changes to `Dockerfile`, `Dockerfile-dev`, and `docker/dev/run.sh`. I guess it is especially important to ensure that the `react-components` will be built when we release the staging/production Portal. 😉

To achieve this:
1. I added a modern Node version to our Docker image. Version 18.18 was the latest one I could make work with the Alpine Linux we use. It's sufficient, and I didn't want to spend more time on this, as I wasn't sure why we use that specific Alpine image version.
2. I removed the checked-in `react-components.js` and `react-test-globals.js` files, and added a git-ignored `app/assets/javascripts/builds` directory where webpack builds the react-components.
3. I moved `react-components` to `rails/react-component` so the source is available in the Docker image and can be built automatically.
4. I updated the `Dockerfile` and `docker/dev/run.sh` to build react-components automatically.
5. I updated GitHub CI to lint, build, and test `react-components`.
6. I removed the empty lint job from GitHub CI (unrelated to other work, but since I was there).

Now, a developer who works with Portal and simply runs `docker-compose up` doesn't need to do anything else if they are working on Rails code. The react-components will be built automatically at the beginning.

I perform the build just once in `docker/dev/run.sh`, without starting `webpack --watch`, as that would require using foreman to run two processes simultaneously -> rails server and `webpack --watch`. That's what jsbundling-rails does and it's not very complicated, but I think it’s more convenient for a developer working on react-components to open a new terminal and run `webpack --watch` separately from all the logs coming from Rails, SOLR, ElasticSearch, etc.

I haven't yet tried to extract our assets like CSS or images from the final JS bundle. This works as it used to – absolutely everything is inlined, so there's no regression. I'll try to work on that tomorrow. However, I wanted to open this PR first, as I hope this is a significant enough improvement and is less likely to break things than the CSS and images extraction.

### CODE REVIEW

File: rails/Dockerfile [hunk 1]
```diff
@@ -56,6 +66,9 @@ RUN \
   && cp ./config/app_environment_variables.sample.rb app_environment_variables.rb \
   && cp -r ./docker/prod/config/initializers/ ./config/
 
+# Change to the react-components directory and run npm build BEFORE rails assets precompile
+RUN cd react-components && npm install && npm run build && cd ..
```
Comment (dougmartin) [inline rails/Dockerfile:70]: Shouldn't this use `npm ci` instead of `npm install` so that it uses the lockfile?

File: rails/docker/dev/run.sh [hunk 2]
```diff
@@ -39,7 +39,12 @@ if [ ! -f $DB_CONFIG ]; then
 fi
 
 if [ "$RAILS_ENV" = "production" ]; then
+  # Build react components, prod version
+  cd react-components && npm install && npm run build && cd ..
```
Comment (dougmartin) [inline rails/docker/dev/run.sh:43]: Same question here and on line 47 - shouldn't this use `npm ci` instead of `npm install` so that it uses the lockfile?

**Review by dougmartin — APPROVED** [review]
Overall looks great :+1:  I just had one question about using `npm install` vs `npm ci` to ensure consistency via the lockfile.

**Comment by pjanik** [comment]
Thanks @dougmartin, I changed `npm install` to `npm ci`, it makes sense. Another advantage over `jsbundling-rails` gem that uses `npm install` in its rake commands and we wouldn't be able to change it. :)

**Pull request merged**